File: windows-app-cs/MediaCrop/Helpers/VideoProcessor.cs
using System.Diagnostics;
using System.IO;
using System.Text.Json;
using MediaCrop.Models;

namespace MediaCrop.Helpers;

public static class VideoProcessor
{
    public static VideoInfo ProbeVideo(string path)
    {
        var psi = new ProcessStartInfo(
            "ffprobe",
            $"-v quiet -print_format json -show_streams -show_format \"{path}\"")
        {
            RedirectStandardOutput = true,
            RedirectStandardError  = true,
            UseShellExecute        = false,
            CreateNoWindow         = true,
        };

        Process p;
        try { p = Process.Start(psi)!; }
        catch
        {
            throw new Exception(
                "FFmpeg / ffprobe not found in PATH.\n" +
                "Download from https://ffmpeg.org/download.html and add the bin\\ folder to PATH.");
        }

        string jsonStr = p.StandardOutput.ReadToEnd();
        p.WaitForExit(30000);

        JsonElement root;
        try { root = JsonDocument.Parse(jsonStr).RootElement; }
        catch { throw new Exception("ffprobe returned unexpected output."); }

        var stream = root.GetProperty("streams").EnumerateArray()
            .FirstOrDefault(s => s.TryGetProperty("codec_type", out var ct)
                                 && ct.GetString() == "video");

        int rawW  = stream.TryGetProperty("width",  out var w) ? w.GetInt32()  : 0;
        int rawH  = stream.TryGetProperty("height", out var h) ? h.GetInt32()  : 0;
        double durS = 0;
        if (root.TryGetProperty("format", out var fmt) &&
            fmt.TryGetProperty("duration", out var d))
            double.TryParse(d.GetString(), System.Globalization.NumberStyles.Float,
                System.Globalization.CultureInfo.InvariantCulture, out durS);

        int rotation = 0;
        if (stream.TryGetProperty("side_data_list", out var sdl))
        {
            foreach (var sd in sdl.EnumerateArray())
            {
                if (sd.TryGetProperty("side_data_type", out var sdt) &&
                    sdt.GetString() == "Display Matrix" &&
                    sd.TryGetProperty("rotation", out var rot))
                {
                    rotation = ((-rot.GetInt32()) % 360 + 360) % 360;
                    break;
                }
            }
        }
        if (rotation == 0 && stream.TryGetProperty("tags", out var tags) &&
            tags.TryGetProperty("rotate", out var r))
        {
            int.TryParse(r.GetString(), out rotation);
            rotation = (rotation % 360 + 360) % 360;
        }

        return new VideoInfo
        {
            Path       = path,
            RawWidth   = rawW,
            RawHeight  = rawH,
            DurationMs = (long)(durS * 1000),
            Rotation   = rotation,
        };
    }

    public static async Task ExportAsync(
        VideoInfo info,
        CropRect  crop,
        long      startMs,
        long      endMs,
        string    outputPath,
        IProgress<double> progress,
        CancellationToken ct)
    {
        long dur = endMs - startMs;
        int dw = info.DisplayWidth, dh = info.DisplayHeight;

        int cx = (int)(crop.Left   * dw);
        int cy = (int)(crop.Top    * dh);
        int cw = (int)(crop.Width  * dw); cw -= cw % 2; cw = Math.Max(2, cw);
        int ch = (int)(crop.Height * dh); ch -= ch % 2; ch = Math.Max(2, ch);

        string cropFilter = $"crop={cw}:{ch}:{cx}:{cy}";
        string vf = info.Rotation switch
        {
            90  => $"transpose=1,{cropFilter}",
            180 => $"hflip,vflip,{cropFilter}",
            270 => $"transpose=2,{cropFilter}",
            _   => cropFilter,
        };

        string args =
            $"-y -ss {startMs / 1000.0:F3} -i \"{info.Path}\" -t {dur / 1000.0:F3} " +
            $"-vf \"{vf}\" -c:v libx264 -crf 18 -preset fast -movflags +faststart " +
            $"-an -metadata:s:v:0 rotate=0 -progress pipe:1 \"{outputPath}\"";

        var psi = new ProcessStartInfo("ffmpeg", args)
        {
            RedirectStandardOutput = true,
            RedirectStandardError  = true,
            UseShellExecute        = false,
            CreateNoWindow         = true,
        };

        Process proc;
        try { proc = Process.Start(psi)!; }
        catch { throw new Exception("FFmpeg not found in PATH."); }

        string? line;
        while ((line = await proc.StandardOutput.ReadLineAsync(ct)) != null)
        {
            ct.ThrowIfCancellationRequested();
            if (line.StartsWith("out_time_us=") &&
                long.TryParse(line["out_time_us=".Length..], out long us) && dur > 0)
            {
                progress.Report(Math.Min((double)us / (dur * 1000.0), 0.99));
            }
        }

        await proc.WaitForExitAsync(ct);
        if (proc.ExitCode != 0)
        {
            string err = await proc.StandardError.ReadToEndAsync(ct);
            int cut = Math.Max(0, err.Length - 600);
            throw new Exception($"FFmpeg failed (exit {proc.ExitCode}):\n{err[cut..]}");
        }

        progress.Report(1.0);
    }
}

File: windows-app/app/video_processor.py
import subprocess
import json
import re
from pathlib import Path
from PyQt6.QtCore import QThread, pyqtSignal
from .models import VideoInfo, CropRect


def probe_video(path: str) -> VideoInfo:
    """Run ffprobe and return VideoInfo.  Raises RuntimeError if ffprobe not found."""
    cmd = [
        "ffprobe", "-v", "quiet",
        "-print_format", "json",
        "-show_streams", "-show_format",
        path,
    ]
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=30)
    except FileNotFoundError:
        raise RuntimeError(
            "FFmpeg / ffprobe not found in PATH.\n"
            "Download FFmpeg from https://ffmpeg.org/download.html and add the bin folder to PATH."
        )

    try:
        data = json.loads(result.stdout)
    except json.JSONDecodeError:
        raise RuntimeError(f"ffprobe returned unexpected output:\n{result.stderr[:400]}")

    stream = next(
        (s for s in data.get("streams", []) if s.get("codec_type") == "video"),
        None,
    )
    if stream is None:
        raise ValueError("No video stream found in this file.")

    width  = int(stream.get("width",  0))
    height = int(stream.get("height", 0))
    dur_s  = float(data.get("format", {}).get("duration", 0))

    # Rotation — stored as side_data Display Matrix or as a tags entry
    rotation = 0
    for sd in stream.get("side_data_list", []):
        if sd.get("side_data_type") == "Display Matrix":
            try:
                rotation = (-int(sd.get("rotation", 0))) % 360
            except (TypeError, ValueError):
                pass
            break
    if rotation == 0:
        try:
            rotation = int(stream.get("tags", {}).get("rotate", 0)) % 360
        except (TypeError, ValueError):
            pass

    return VideoInfo(
        path=path,
        raw_width=width,
        raw_height=height,
        duration_ms=int(dur_s * 1000),
        rotation=rotation,
    )


class ExportThread(QThread):
    progress  = pyqtSignal(float)   # 0.0–1.0
    completed = pyqtSignal(str)     # output file path
    failed    = pyqtSignal(str)     # error message

    def __init__(
        self,
        video_info: VideoInfo,
        crop_rect:  CropRect,
        start_ms:   int,
        end_ms:     int,
        output_path: str,
        parent=None,
    ):
        super().__init__(parent)
        self._vi   = video_info
        self._crop = crop_rect
        self._start_ms  = start_ms
        self._end_ms    = end_ms
        self._output    = output_path

    def run(self):
        try:
            self._export()
        except Exception as exc:
            self.failed.emit(str(exc))

    def _export(self):
        vi   = self._vi
        crop = self._crop
        dur  = self._end_ms - self._start_ms

        dw = vi.display_width
        dh = vi.display_height

        # Crop region in display-space pixels, rounded to even
        cx = int(crop.left   * dw)
        cy = int(crop.top    * dh)
        cw = int(crop.width  * dw); cw -= cw % 2
        ch = int(crop.height * dh); ch -= ch % 2
        cw = max(2, cw); ch = max(2, ch)

        # Build video filter: rotate first so crop operates in display coords
        crop_f = f"crop={cw}:{ch}:{cx}:{cy}"
        rot = vi.rotation
        if   rot == 90:  vf = f"transpose=1,{crop_f}"
        elif rot == 180: vf = f"hflip,vflip,{crop_f}"
        elif rot == 270: vf = f"transpose=2,{crop_f}"
        else:            vf = crop_f

        cmd = [
            "ffmpeg", "-y",
            "-ss", f"{self._start_ms / 1000:.3f}",
            "-i", vi.path,
            "-t",  f"{dur / 1000:.3f}",
            "-vf", vf,
            "-c:v", "libx264",
            "-crf", "18",
            "-preset", "fast",
            "-movflags", "+faststart",
            "-an",                          # always muted — by design
            "-metadata:s:v:0", "rotate=0",  # clear rotation tag (already applied)
            "-progress", "pipe:1",
            self._output,
        ]

        try:
            proc = subprocess.Popen(
                cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
            )
        except FileNotFoundError:
            raise RuntimeError(
                "FFmpeg not found in PATH.\n"
                "Download FFmpeg from https://ffmpeg.org/download.html and add the bin folder to PATH."
            )

        for line in proc.stdout:
            line = line.strip()
            if line.startswith("out_time_us="):
                try:
                    us = int(line.split("=", 1)[1])
                    if dur > 0:
                        self.progress.emit(min(us / (dur * 1000), 0.99))
                except (ValueError, ZeroDivisionError):
                    pass

        proc.wait()
        if proc.returncode != 0:
            stderr = proc.stderr.read() if proc.stderr else ""
            raise RuntimeError(f"FFmpeg failed (exit {proc.returncode}):\n{stderr[-600:]}")

        self.progress.emit(1.0)
        self.completed.emit(self._output)
